File: apps/backend/app/schemas/knowledge_graph.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.knowledge_graph import (
    EntityType,
    RelationType,
)


class KnowledgeGraphNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    entity_type: EntityType
    name: str
    description: str | None = None
    source_document: str | None = None
    created_at: datetime


class KnowledgeGraphEdge(BaseModel):
    id: int
    source: int
    target: int
    relation_type: RelationType


class KnowledgeGraphResponse(BaseModel):
    project_id: int
    nodes: list[KnowledgeGraphNode]
    edges: list[KnowledgeGraphEdge]


class EntityDetailResponse(KnowledgeGraphNode):
    related_entities: list[KnowledgeGraphNode]

File: apps/frontend/types/knowledge_graph.ts
export type EntityType =
  | "substation"
  | "bay"
  | "panel"
  | "circuit_breaker"
  | "disconnector"
  | "transformer"
  | "current_transformer"
  | "voltage_transformer"
  | "protection_relay"
  | "cable"
  | "signal"
  | "document"
  | "test"
  | "other";

export type RelationType =
  | "belongs_to"
  | "connected_to"
  | "protected_by"
  | "documented_in"
  | "tested_by"
  | "part_of"
  | "other";

export interface KnowledgeGraphNode {
  id: number;
  project_id: number;
  entity_type: EntityType;
  name: string;
  description: string | null;
  source_document: string | null;
  created_at: string;
}

export interface KnowledgeGraphEdge {
  id: number;
  source: number;
  target: number;
  relation_type: RelationType;
}

export interface KnowledgeGraph {
  project_id: number;
  nodes: KnowledgeGraphNode[];
  edges: KnowledgeGraphEdge[];
}